### DISCUSSION TITLE
`magit-diff-buffer-file-unstaged` ?

### DISCUSSION THREAD

**Discussion by elibarzilay** [magit/magit#5013] [Ideas]
One of the fundamentally useful things about magit is staging random bits on the status screen.  But sometimes it's more convenient to work inside a file -- and then I have `magit-diff-buffer-file` as a quick way to see what I did.  But I don't see any convenient way to get an `-unstaged` version of that, showing just these changes (which satisfyingly disappear once I `s`tage them).

TLDR: It would be nice if there was some `magit-diff-buffer-file-unstaged`, which is just like the existing `magit-diff-buffer-file`, except passing `nil` as the first argument to `magit-diff-setup-buffer`, which means that a weak version of implementing this feature could be just a quick new optional argument to the function, which is what I've hacked for now:

```
(defun magit-diff-buffer-file (&optional mode)
  ...
              (magit-diff-setup-buffer (and (not (eq mode 'unstaged))
                                            (or (magit-get-current-branch) "HEAD"))
```

But I'm guessing that it's likely that I'm missing something obvious...  (I *did* search a bunch of stuff...)

(Or if I'm not, then this might be part of #2952 ?)

**Comment by tarsius** [comment]
Sounds reasonable but I'll have to experiment with it a bit and I don't think I'll get to that soon.

Meanwhile, use your patch and share any insights you gain from doing that.

**Reply by elibarzilay** [reply]
@tarsius I've been using this sporadically for a while, and didn't encounter any issues.

**Reply by tarsius** [reply]
I was just considering to at least add a prefix argument, to limit to just the unstaged changes, but I concluded that wasn't necessary.

I assume you would want to add a binding for the new command to `magit-file-dispatch` (`C-c f`). `u` is already taken in there, so we would have to, e.g., change `d` to `d d` and bind `d u` to the new command. If we used a prefix argument instead we could leave `d` alone and you would have to use `C-c f C-u d` to show just the unstaged changes.

However there is already `C-c D`, which gives more control over what diff to show and how. That is bound to the regular `magit-diff` menu, but when invoked this way, it is already limited to the visited file. So the existing binding for showing just the unstaged changes for the current file is `C-c f D u`, which is a bit worse than either `C-c f C-u d` or `C-c f d d`, but has the advantage that many other variants are available (instead of the final `u`) and that you can set arguments if desired.

**Reply by tarsius** [reply]
> I was just considering to at least add a prefix argument, to limit to just the unstaged changes, but I concluded that wasn't necessary.

Unnecessary, but adding another road to Rome doesn't hurt. Implemented in 5429811150cc487603eaae972c0e4937b16651ac.

**Reply by elibarzilay** [reply]
@tarsius Thanks!!  And apologies for not getting back to it back then...

The thing is that I was fully expecting to look deeper into into it and find something existing. In any case, I indeed settled on `magit-file-dispatch` with

    (transient-insert-suffix 'magit-file-dispatch "d"
      `("x" "Diff Unstaged"
        ,(lambda ()
           (interactive)
           (magit-diff-buffer-file 'unstaged) ; <- my hacked version
           (delete-other-windows))))

But you're right that `D u` is pretty much the same (the only visible difference I see is a split window next to the file vs a whole window, but that might be my overcomplicated setup).

The main thing is that since staging/unstaging is such a useful thing in the status, that I see a plain `d` is nearly useless since `s` or `u` things doesn't translate to any form of feedback for a view that shows both staged and unstaged hunks...
